File: frontend/src/types/marketing.ts
// Layout JSON structures

export type BlockType = 'logo' | 'text' | 'qr' | 'image' | 'shape'

export interface LayoutBlock {
  id: string
  type: BlockType
  binding?: string | null // e.g., business.name, business.qr.profile
  x: number
  y: number
  width?: number
  height?: number
  size?: number // For QR codes
  unit: string
  editable_by_business: boolean
  editable_by_support: boolean
  // Text-specific
  text?: string
  fontFamily?: string
  fontSizePt?: number
  fontWeight?: string
  align?: 'left' | 'center' | 'right'
  color?: string
  // QR-specific
  qr_url?: string
}

export interface LayoutPaper {
  size: 'A3' | 'A4' | 'A5'
  orientation: 'portrait' | 'landscape'
  width_mm: number
  height_mm: number
}

export interface LayoutTheme {
  background: string
  primaryColor: string
  accentColor: string
}

export interface LayoutJson {
  version: number
  paper: LayoutPaper
  theme: LayoutTheme
  blocks: LayoutBlock[]
}

// Marketing Templates (global blueprints)

export interface MarketingTemplate {
  id: string
  template_key: string
  name: string
  description?: string | null
  paper_size: 'A3' | 'A4' | 'A5'
  orientation: 'portrait' | 'landscape'
  layout_schema_version: number
  layout_json: LayoutJson
  thumbnail_url?: string | null
  is_active: boolean
  sort_order: number
  created_at: string
  updated_at: string
}

export interface MarketingTemplatesResponse {
  items: MarketingTemplate[]
  total: number
}

// Marketing Materials (business instances)

export interface MarketingMaterial {
  id: string
  business_id: string
  template_id?: string | null
  name: string
  paper_size: 'A3' | 'A4' | 'A5'
  orientation: 'portrait' | 'landscape'
  layout_schema_version: number
  layout_json: LayoutJson
  is_default_for_business: boolean
  support_notes?: string | null
  created_by_user_id?: string | null
  updated_by_user_id?: string | null
  created_at: string
  updated_at: string
}

export interface MarketingMaterialsResponse {
  items: MarketingMaterial[]
  total: number
}

export interface MarketingMaterialCreate {
  template_id: string
  name?: string
}

export interface MarketingMaterialUpdate {
  name?: string
  layout_json?: LayoutJson
}

File: backend/app/schemas/marketing.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class LayoutBlock(BaseModel):
    """Individual block in layout JSON."""
    id: str
    type: Literal['logo', 'text', 'qr', 'image', 'shape']
    binding: str | None = None  # e.g., business.name, business.qr.profile
    x: float
    y: float
    width: float | None = None
    height: float | None = None
    size: float | None = None  # For QR codes
    unit: str = 'mm'
    editable_by_business: bool = True
    editable_by_support: bool = True
    # Text-specific
    text: str | None = None
    fontFamily: str | None = None
    fontSizePt: float | None = None
    fontWeight: str | None = None
    align: str | None = None
    color: str | None = None
    # QR-specific
    qr_url: str | None = None


class LayoutPaper(BaseModel):
    """Paper settings in layout."""
    size: str = 'A4'
    orientation: str = 'portrait'
    width_mm: float
    height_mm: float


class LayoutTheme(BaseModel):
    """Theme settings in layout."""
    background: str = '#FFFFFF'
    primaryColor: str = '#1a1a1a'
    accentColor: str = '#2563eb'


class LayoutJson(BaseModel):
    """Full layout JSON structure."""
    version: int = 1
    paper: LayoutPaper
    theme: LayoutTheme
    blocks: list[LayoutBlock] = Field(default_factory=list)


# ============================================
# Marketing Templates (global)
# ============================================

class MarketingTemplateBase(BaseModel):
    template_key: str
    name: str
    description: str | None = None
    paper_size: Literal['A3', 'A4', 'A5'] = 'A4'
    orientation: Literal['portrait', 'landscape'] = 'portrait'


class MarketingTemplate(MarketingTemplateBase):
    id: str
    layout_schema_version: int = 1
    layout_json: dict[str, Any]
    thumbnail_url: str | None = None
    is_active: bool = True
    sort_order: int = 0
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class MarketingTemplatesResponse(BaseModel):
    items: list[MarketingTemplate]
    total: int


# ============================================
# Marketing Materials (business instances)
# ============================================

class MarketingMaterialBase(BaseModel):
    name: str
    paper_size: Literal['A3', 'A4', 'A5'] = 'A4'
    orientation: Literal['portrait', 'landscape'] = 'portrait'


class MarketingMaterialCreate(BaseModel):
    template_id: str
    name: str | None = None  # Optional, defaults to template name


class MarketingMaterialUpdate(BaseModel):
    name: str | None = None
    layout_json: dict[str, Any] | None = None


class MarketingMaterialAdminUpdate(MarketingMaterialUpdate):
    """Admin/support can also update support_notes."""
    support_notes: str | None = None


class MarketingMaterial(MarketingMaterialBase):
    id: str
    business_id: str
    template_id: str | None = None
    layout_schema_version: int = 1
    layout_json: dict[str, Any]
    is_default_for_business: bool = False
    support_notes: str | None = None
    created_by_user_id: str | None = None
    updated_by_user_id: str | None = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class MarketingMaterialWithTemplate(MarketingMaterial):
    """Material with template info for display."""
    template: MarketingTemplate | None = None


class MarketingMaterialsResponse(BaseModel):
    items: list[MarketingMaterial]
    total: int
